sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala

```diff
@@ -217,3 +217,14 @@ private[sql] abstract class FractionalType extends NumericType {
   private[sql] val fractional: Fractional[InternalType]
   private[sql] val asIntegral: Integral[InternalType]
 }
+
+private[sql] object AnyTimestampType extends AbstractDataType with Serializable {
```

Comment:
Can we use a type collection like `TypeCollection.NumericAndInterval`?

Reply:
I do have introduced `TypeCollection.AllTimestampTypes` in a previous PR.
However, in this PR I find that having `AnyTimestampType` can simplify the code changes in `TypeCoercion.scala`.

Reply:
Are we going to use `AnyTimestampType` everywhere and remove `AllTimestampTypes`?

Reply:
Yes, I will have a follow-up for that.
There is another reason, in the default type coercion rules, when implicit casting a type to a `TypeCollection` type, Spark chooses the first convertible data type as the result. If we are going to make the default timestamp type configurable, having `AnyTimestampType` is better.

Reply:
So we should have such a new type for types of the same category, like `NumericType` and `IntegralType`.
If the data types are of different category, it's fine to use `TypeCollection`